diff --git a/cards/delphes_card_ATLAS.tcl b/cards/delphes_card_ATLAS.tcl
index 0d7efb4..4723d5c 100644
--- a/cards/delphes_card_ATLAS.tcl
+++ b/cards/delphes_card_ATLAS.tcl
@@ -462,7 +462,7 @@ module Isolation PhotonIsolation {
 
   set OutputArray photons
 
-  set DeltaRMax 0.5
+  set DeltaRMax 0.4
 
   set PTMin 0.5
 
@@ -497,7 +497,7 @@ module Isolation ElectronIsolation {
 
   set OutputArray electrons
 
-  set DeltaRMax 0.5
+  set DeltaRMax 0.4
 
   set PTMin 0.5
 
@@ -531,7 +531,7 @@ module Isolation MuonIsolation {
 
   set OutputArray muons
 
-  set DeltaRMax 0.5
+  set DeltaRMax 0.4
 
   set PTMin 0.5
 
@@ -593,7 +593,7 @@ module FastJetFinder GenJetFinder {
 
   # algorithm: 1 CDFJetClu, 2 MidPoint, 3 SIScone, 4 kt, 5 Cambridge/Aachen, 6 antikt
   set JetAlgorithm 6
-  set ParameterR 0.6
+  set ParameterR 0.4
 
   set JetPTMin 20.0
 }
@@ -622,7 +622,7 @@ module FastJetFinder FastJetFinder {
 
   # algorithm: 1 CDFJetClu, 2 MidPoint, 3 SIScone, 4 kt, 5 Cambridge/Aachen, 6 antikt
   set JetAlgorithm 6
-  set ParameterR 0.6
+  set ParameterR 0.4
 
   set JetPTMin 20.0
 }
@@ -650,7 +650,7 @@ module JetFlavorAssociation JetFlavorAssociation {
   set ParticleLHEFInputArray Delphes/allParticlesLHEF
   set JetInputArray JetEnergyScale/jets
 
-  set DeltaR 0.5
+  set DeltaR 0.4
   set PartonPTMin 1.0
   set PartonEtaMax 2.5
 
diff --git a/cards/delphes_card_ATLAS_PileUp.tcl b/cards/delphes_card_ATLAS_PileUp.tcl
index 171fc8b..9cad29f 100644
--- a/cards/delphes_card_ATLAS_PileUp.tcl
+++ b/cards/delphes_card_ATLAS_PileUp.tcl
@@ -45,17 +45,19 @@ set ExecutionPath {
   
   MissingET
   
-  JetFlavorAssociation
-
-  BTagging
-  TauTagging
-
   UniqueObjectFinder
 
   ScalarHT
 
+  BSMFilter
   TreeWriter
 }
+#  JetFlavorAssociation
+#  JetFlavorAssociationGen
+#  BTagging
+#  TauTagging
+#  UniqueObjectFinderPUjets
+
 
 ###############
 # PileUp Merger
@@ -68,10 +70,11 @@ module PileUpMerger PileUpMerger {
   set VertexOutputArray vertices
 
   # pre-generated minbias input file
-  set PileUpFile MinBias.pileup
+  set PileUpFile /afs/cern.ch/work/j/jmontejo/LLPtriggertiming/LLPtrigger_samples/Minbias.pileup
 
   # average expected pile up
-  set MeanPileUp 50
+  set MeanPileUp 55
+  #set PileUpDistribution 2 #0 Poisson (default), 1 Uniform, 2 Fixed value
 
    # maximum spread in the beam direction in m
   set ZVertexSpread 0.25
@@ -313,6 +316,14 @@ module Calorimeter Calorimeter {
 # Electron filter
 #################
 
+module PdgCodeFilter BSMFilter {
+  set InputArray Delphes/allParticles
+  set OutputArray bsmParticles
+  set Invert true
+  add PdgCode {1000025}
+  add PdgCode {1000035}
+}
+
 module PdgCodeFilter ElectronFilter {
   set InputArray Calorimeter/eflowTracks
   set OutputArray electrons
@@ -439,7 +450,7 @@ module FastJetFinder GenJetFinder {
 
   # algorithm: 1 CDFJetClu, 2 MidPoint, 3 SIScone, 4 kt, 5 Cambridge/Aachen, 6 antikt
   set JetAlgorithm 6
-  set ParameterR 0.6
+  set ParameterR 0.4
 
   set JetPTMin 20.0
 }
@@ -469,7 +480,7 @@ module FastJetFinder FastJetFinder {
 
   # jet algorithm: 1 CDFJetClu, 2 MidPoint, 3 SIScone, 4 kt, 5 Cambridge/Aachen, 6 antikt
   set JetAlgorithm 6
-  set ParameterR 0.6
+  set ParameterR 0.4
 
   set JetPTMin 20.0
 }
@@ -484,7 +495,7 @@ module JetPileUpSubtractor JetPileUpSubtractor {
 
   set OutputArray jets
 
-  set JetPTMin 20.0
+  set JetPTMin 10.0
 }
 
 ##################
@@ -632,6 +643,19 @@ module Merger ScalarHT {
 # Jet Flavor Association
 ########################
 
+module JetFlavorAssociation JetFlavorAssociationGen {
+  
+  set PartonInputArray Delphes/partons
+  set ParticleInputArray Delphes/allParticles
+  set ParticleLHEFInputArray Delphes/allParticlesLHEF
+  set JetInputArray GenJetFinder/jets
+  
+  set DeltaR 0.4
+  set PartonPTMin 5.0
+  set PartonEtaMax 4.5
+
+}
+
 module JetFlavorAssociation JetFlavorAssociation {
   
   set PartonInputArray Delphes/partons
@@ -639,9 +663,9 @@ module JetFlavorAssociation JetFlavorAssociation {
   set ParticleLHEFInputArray Delphes/allParticlesLHEF
   set JetInputArray JetEnergyScale/jets
   
-  set DeltaR 0.5
-  set PartonPTMin 1.0
-  set PartonEtaMax 2.5
+  set DeltaR 0.4
+  set PartonPTMin 5.0
+  set PartonEtaMax 4.5
 
 }
 
@@ -719,6 +743,15 @@ module UniqueObjectFinder UniqueObjectFinder {
   add InputArray JetEnergyScale/jets jets
 }
 
+module UniqueObjectFinder UniqueObjectFinderPUjets {
+# earlier arrays take precedence over later ones
+# add InputArray InputArray OutputArray
+  add InputArray PhotonIsolation/photons photons
+  add InputArray ElectronIsolation/electrons electrons
+  add InputArray MuonIsolation/muons muons
+  add InputArray FastJetFinder/jets jets
+}
+
 ##################
 # ROOT tree writer
 ##################
@@ -729,10 +762,11 @@ module UniqueObjectFinder UniqueObjectFinder {
 
 module TreeWriter TreeWriter {
 # add Branch InputArray BranchName BranchClass
-  add Branch Delphes/allParticles Particle GenParticle
+  add Branch BSMFilter/bsmParticles Particle GenParticle
+#  add Branch Delphes/allParticles ParticleAll GenParticle
 
 #  add Branch TrackMerger/tracks Track Track
-  add Branch Calorimeter/towers Tower Tower
+#  add Branch Calorimeter/towers Tower Tower
 
 #  add Branch Calorimeter/eflowTracks EFlowTrack Track
 #  add Branch Calorimeter/eflowPhotons EFlowPhoton Tower
@@ -741,6 +775,7 @@ module TreeWriter TreeWriter {
   add Branch GenJetFinder/jets GenJet Jet
   add Branch GenMissingET/momentum GenMissingET MissingET
 
+#  add Branch UniqueObjectFinderPUjets/jets JetNoPUcorr Jet
   add Branch UniqueObjectFinder/jets Jet Jet
   add Branch UniqueObjectFinder/electrons Electron Electron
   add Branch UniqueObjectFinder/photons Photon Photon
diff --git a/examples/Pythia8/configLHE.cmnd b/examples/Pythia8/configLHE.cmnd
index a301ff8..98762e0 100644
--- a/examples/Pythia8/configLHE.cmnd
+++ b/examples/Pythia8/configLHE.cmnd
@@ -1,6 +1,6 @@
 ! 1) Settings used in the main program.
 
-Main:numberOfEvents = 10         ! number of events to generate
+Main:numberOfEvents = 20000        ! number of events to generate
 Main:timesAllowErrors = 3          ! how many aborts before run stops
 
 ! 2) Settings related to output in init(), next() and stat().
@@ -15,4 +15,4 @@ Next:numberShowEvent = 0           ! print event record n times
 ! 3) Set the input LHE file
 
 Beams:frameType = 4
-Beams:LHEF = examples/Pythia8/events.lhe    
+Beams:LHEF = /afs/cern.ch/work/j/jmontejo/LLPtriggertiming/LLPtrigger_samples/n3n4_1TeVsquark_LHE_events.150gev.events
diff --git a/examples/Pythia8/configNoLHE.cmnd b/examples/Pythia8/configNoLHE.cmnd
index c1fe33d..0399ced 100644
--- a/examples/Pythia8/configNoLHE.cmnd
+++ b/examples/Pythia8/configNoLHE.cmnd
@@ -13,12 +13,17 @@ Next:numberShowEvent = 0           ! print event record n times
 ! 3) Beam parameter settings. Values below agree with default ones.
 Beams:idA = 2212                   ! first beam, p = 2212, pbar = -2212
 Beams:idB = 2212                   ! second beam, p = 2212, pbar = -2212
-Beams:eCM = 14000.                 ! CM energy of collision
+Beams:eCM = 13000.                 ! CM energy of collision
 
 ! 4) Settings for the hard-process generation.
 
-Top:gg2ttbar = on                  ! g g -> t tbar
-Top:qqbar2ttbar = on               ! q qbar -> t tbar
+SoftQCD:all = on
+#HardQCD:all = on
+#PhaseSpace:Bias2Selection=on
+#PhaseSpace:pTHatMin = 8.
+
+#Top:gg2ttbar = on                  ! g g -> t tbar
+#Top:qqbar2ttbar = on               ! q qbar -> t tbar
 
 ! 5) Switch on/off the key event generation steps.
 #PartonLevel:MPI = off              ! no multiparton interactions
diff --git a/examples/Pythia8/generatePileUp.cmnd b/examples/Pythia8/generatePileUp.cmnd
index be142dd..9cb4aae 100644
--- a/examples/Pythia8/generatePileUp.cmnd
+++ b/examples/Pythia8/generatePileUp.cmnd
@@ -5,7 +5,7 @@
 ! The changes here are illustrative, not always physics-motivated.
 
 ! 1) Settings that will be used in a main program.
-Main:numberOfEvents = 1000          ! number of events to generate
+Main:numberOfEvents = 100          ! number of events to generate
 Main:timesAllowErrors = 3          ! abort run after this many flawed events
 
 ! 2) Settings related to output in init(), next() and stat().
@@ -13,7 +13,7 @@ Init:showChangedSettings = on      ! list changed settings
 Init:showAllSettings = off         ! list all settings
 Init:showChangedParticleData = on  ! list changed particle data
 Init:showAllParticleData = off     ! list all particle data
-Next:numberCount = 5            ! print message every n events
+Next:numberCount = 100             ! print message every n events
 Next:numberShowLHA = 1             ! print LHA information n times
 Next:numberShowInfo = 1            ! print event information n times
 Next:numberShowProcess = 1         ! print process record n times
diff --git a/modules/JetFlavorAssociation.cc b/modules/JetFlavorAssociation.cc
index aae3b3a..11b0833 100644
--- a/modules/JetFlavorAssociation.cc
+++ b/modules/JetFlavorAssociation.cc
@@ -217,6 +217,21 @@ void JetFlavorAssociation::Process()
 // Standard definition of jet flavor in
 // https://cmssdt.cern.ch/SDT/lxr/source/PhysicsTools/JetMCAlgos/plugins/JetPartonMatcher.cc?v=CMSSW_7_3_0_pre1
 
+Candidate * getLastParton(Candidate* parton, const TObjArray *theArray){
+    Candidate *tmp;
+    for(int id_child=parton->D1; id_child <= parton->D2; id_child++){
+        tmp = static_cast<Candidate *>(theArray->At(id_child));
+        //if (parton->PID == 3)
+        //std::cout << "id,in,next PID, status:" << id_child << " " << parton->PID << " " << tmp->PID << " " << tmp->Status << std::endl;
+
+        if (tmp->PID == parton->PID){
+            parton = getLastParton(tmp, theArray);
+            break;
+        }
+    }
+    return parton;
+}
+
 void JetFlavorAssociation::GetAlgoFlavor(Candidate *jet, TObjArray *partonArray, TObjArray *partonLHEFArray)
 {
   float maxPt = 0;
@@ -233,6 +248,8 @@ void JetFlavorAssociation::GetAlgoFlavor(Candidate *jet, TObjArray *partonArray,
   {
     // default delphes method
     pdgCode = TMath::Abs(parton->PID);
+    if(parton->Status != 23) continue; //JAVIER my change
+    //parton = getLastParton(parton, fParticleInputArray);
     if(TMath::Abs(parton->PID) == 21) pdgCode = 0;
     if(jet->Momentum.DeltaR(parton->Momentum) <= fDeltaR)
     {
